Implement atomic modesetting in DRM

no92 opened the pull request:
This PR aims to land atomic modesetting support in DRM and wiring up `vmware` and `plainfb` drivers for it. Atomic modesetting has been tested with `weston` and the [`drm-howto`](https://github.com/dvdhrm/docs/). `kmscon` does not utilize atomic modesetting.

These changes depend on managarm/mlibc#272.

no92 commented:
Added DRM states for better atomic modesetting and avoiding data duplication, should be ready for review now.

avdgrinten commented on core/drm/include/core/drm/core.hpp:
```diff
@@ -225,6 +312,9 @@ struct File {
 
 	protocols::fs::StatusPageProvider _statusPage;
+
+	bool universal_planes;
```
camelCase.

avdgrinten commented on core/drm/include/core/drm/core.hpp:
```diff
@@ -52,28 +52,70 @@ struct BlobPropertyType {
+	uint32_t flags();
 	PropertyType propertyType();
+	std::string name();
+	void add_enum_info(uint64_t value, std::string name);
+	const std::unordered_map<uint64_t, std::string>& enum_info();
```
What do these methods do?

no92 replied:
Querying property values returns an integer value for enums, where associating that numeric value with a enums member requires retrieving the mapping of numeric value to enum. These methods are for adding and retrieving those mappings.

avdgrinten reviewed (commented):
Quick review of the first two commits only. I still have to look at the new infrastructure in detail.

avdgrinten commented on core/drm/include/core/drm/core.hpp:
```diff
@@ -262,10 +358,16 @@ struct ModeObject {
 	void setupWeakPtr(std::weak_ptr<ModeObject> self);
 	std::shared_ptr<ModeObject> sharedModeObject();
 
+	bool setAssignment(const Assignment a);
```
How does `setAssignment` interact with `Configuration`? As far as I can see it does not, at least not in the first commit?

no92 replied:
`setAssignment` is a relic of the old way of setting properties and gets removed in later commits.

no92 commented on core/drm/include/core/drm/core.hpp:
```diff
@@ -52,28 +52,70 @@ struct BlobPropertyType {
 	uint64_t timestamp;
 };
 
+enum PropertyId {
+	Invalid, /* 0 */
+	SrcW,
+	SrcH,
+	FbId,
+	ModeId,
+	CrtcX, /* 5 */
+	CrtcY,
+	PlaneType,
+	Dpms,
+	CrtcId,
+	Active, /* 10 */
+	SrcX,
+	SrcY,
+	CrtcW,
+	CrtcH,
+};
```
This served debugging purposes when logging DRM requests to decode property IDs and is probably unnecessary now, as property names can be retrieved by calling `name()` on the property.

no92 commented:
Addressed the issues raised above.

qookei commented on core/drm/src/core.cpp:
```diff
@@ -1003,6 +1550,249 @@ drm_core::File::ioctl(void *object, managarm::fs::CntRequest req,
+		helix::SendBuffer send_resp;
+		managarm::fs::SvrResponse resp;
+
+		if(req.drm_capability() == DRM_CLIENT_CAP_STEREO_3D) {
+			std::cout << "\e[31mcore/drm: DRM client cap for stereo 3D unsupported\e[39m" << std::endl;
```
Missing setting the error code here?

qookei commented on core/drm/src/core.cpp:
```diff
@@ -1003,6 +1550,249 @@ drm_core::File::ioctl(void *object, managarm::fs::CntRequest req,
+		co_await transmit.async_wait();
+		HEL_CHECK(send_resp.error());
+	}else if(req.command() == DRM_IOCTL_MODE_SETPROPERTY) {
+		helix::SendBuffer send_resp;
+		managarm::fs::SvrResponse resp;
+
+		std::cout << "\e[31mcore/drm: DRM_IOCTL_MODE_SETPROPERTY is a no-op\e[39m"
+				<< std::endl;
+
+		auto ser = resp.SerializeAsString();
```
No error code set here either?

qookei commented on core/drm/src/core.cpp:
```diff
@@ -208,19 +367,149 @@ drm_core::Property *drm_core::Device::crtcXProperty() {
 // Property
 // ----------------------------------------------------------------
 
+std::vector<drm_core::Assignment> drm_core::ModeObject::getAssignments(std::shared_ptr<Device> dev) {	switch(this->type()) {
```
Missing newline?

qookei commented on core/drm/src/core.cpp:
```diff
@@ -1003,6 +1550,249 @@ drm_core::File::ioctl(void *object, managarm::fs::CntRequest req,
+			prop_count += req.drm_prop_counts(i);
+		}
+
+		{
+			auto valid = config->capture(assignments, state);
+			assert(valid);
+		}
```
Why is this in it's own block? Isn't `valid` just a bool?

no92 replied:
This is its own block due to the `goto` in line 1744.

qookei reviewed (commented):
Overall seems good, apart from the comments. Also one general thing is that some member variables and methods use snake case and some don't.

no92 commented:
Addressed the issues raised above, removed one unused function and adjusted code style to match the guidelines. However, the accessors for DRM properties are spelled the way as in the original, that is including underscores, but in lower case.

no92 commented:
Removed redundant getters and setters, as discussed on Discord.

avdgrinten commented:
LGTM now, thanks a lot! Atomic modeset enables quite a lot of useful stuff.
